### PR TITLE
Refinement types and unification for symbolic shape inference

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62682
* #62582
* **#61776**

Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

### CODE REVIEW

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 1]
```diff
@@ -500,3 +511,148 @@ def type_check_node(self, n: Node):
...
+
+        for t1, t2 in zip(n.type.__args__[start_dim:end_dim], n.args[0].type.__args__[start_dim:end_dim]):
+            eq_const.append(Equality(t1, t2))
+
+        return eq_const
+
+class Refine:
+    """
+    Symbolic shape inference.
+    Generates constraitns over type variables.
```
Comment (vadimkantorov) [inline torch/fx/experimental/graph_gradual_typechecker.py:581]: typo: constraitns

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 2]
```diff
@@ -6,8 +6,16 @@
 from torch.fx.node import Target, Node
 from torch.nn.modules.batchnorm import BatchNorm2d
 from torch.nn.modules.conv import Conv2d
+from torch.fx.experimental.refinement_types import Equality
+from unification import Var  # type: ignore[import]
+
+def conv3x3(in_planes, out_planes, stride=1, groups=1, dilation=1):
+    """3x3 convolution with padding"""
+    return torch.nn.Conv2d(in_planes, out_planes, kernel_size=3, stride=stride,
+                           padding=dilation, groups=groups, bias=False, dilation=dilation)
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:15]: Is this function used?

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 3]
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
...
             raise NotImplementedError("Method not yet implemented")
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
+def first_one(n: Node):
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        return [Equality(arg_type.__args__[0], n.type.__args__[0])]
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:527]: So this is just saying that the batch size at input needs to match the batch size at output?

Can we encode e.g. that the channel dimensions at input/output need to match the parameter value? Or is that not necessary?

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 4]
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
...
+@register_refinement_rule(torch.nn.AdaptiveAvgPool2d)
+@register_refinement_rule(torch.add)
+@register_refinement_rule(operator.add)
+def all_eq(n: Node):
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        args1 = arg_type.__args__
+        args2 = n.type.__args__
+        return [Equality(args1[i], args2[i]) for i in range(len(args1))]
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:541]: For add, isn't this not true in the presence of broadcasting?

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 5]
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
...
+class Refine:
+    """
+    Symbolic shape inference.
+    Generates constraints over type variables.
+    Currently all constraints are equality constraints.
+    """
+    def __init__(self, traced):
+        self.constraints = []
+        self.traced = traced
+        self.curr_symbol = self.symbol_gen()
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:592]: nit: you can use `self.symbol_iter = itertools.count(start=0, step=1)` and then use `next(self.symbol_iter)` to retrieve the value and advance the iterator

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 6]
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
...
+            new_args = []
+            for a in typ.__args__:
+                if a == Dyn:
+                    new_symbol = Var(self.curr_symbol())
+                    new_args.append(new_symbol)
+                else:
+                    new_args.append(a)
+            return TensorType(tuple(new_args))
+        else:
+            return typ
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:630]: style: Can we make a general recursive map function similar to [map_aggregate](https://github.com/pytorch/pytorch/blob/c30048fccfc5914cb3924d9b2905a5e13b3e0b52/torch/fx/node.py#L537) for doing this type of stuff? That way the code for traversing type expressions and yielding a new one will all be in one place and we can update it once and all of the call-sites will have the new behavior

File: torch/fx/experimental/unify_refinements.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        return TensorType(tuple(new_type))
+
+
+def substitute_all_types(graph, mapping):
+    """
+    Apply the most general unifier to all types in a graph
+    till reaching a fixed point. If the input and output graph
+    are the same, we converge.
+    """
+    old_graph = deepcopy(graph)
```
Comment (jamesr66a) [inline torch/fx/experimental/unify_refinements.py:64]: Any way we could do this in-place? I think this is already going to be very high complexity and having to copy over the entire graph every time may make it even higher

**Review by jamesr66a — COMMENTED** [review]
Looks cool! Just did a quick pass over it, will look deeper into the algorithmic aspects and play around with it tomorrow or wednesday

**Review by jamesr66a — COMMENTED** [review]
BTW can you add a commit message on the PR?

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 3 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:527]: We already have this in the initial pass

```
        w_in = arg_type.__args__[3]
        h_in = arg_type.__args__[2]
        h_out = calculate_out_dimension(h_in, module_instance, 0)
        w_out = calculate_out_dimension(w_in, module_instance, 1)
        new_type = TensorType((arg_type.__args__[0], module_instance.out_channels, h_out, w_out))
```

So I think we already have that part but maybe we consider moving it here at some point so we can do one pass instead of two passes. I haven't worked out what all the constraints would look like in that case though, so I'll think about it.

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 4 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:541]: assuming we run the type inferencer first, broadcasting would have been applied to the types by this point, but maybe if types are not present we don't know if broadcasting happens? I think tomorrow we can discuss this point

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 8]
```diff
@@ -500,3 +522,169 @@ def type_check_node(self, n: Node):
...
+        arg_type1 = n.args[0].type
+        arg_type2 = n.args[0].type
+        if isinstance(arg_type1, TensorType) and isinstance(arg_type2, TensorType) and isinstance(n.type, TensorType):
+            args1 = arg_type1.__args__
+            args2 = arg_type2.__args__
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
+                if args1 == args2:
+                    r.append(Equality(x, z))
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:565]: i think this isnt handling broadcasting

File: torch/fx/experimental/unify_refinements.py (hunk 7 above)
Comment (migeed-z) [reply torch/fx/experimental/unify_refinements.py]: I am doing this in place now with a dictionary. Does that seem fine or can there be a better way to do this?

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 8 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:565]: I updated this part. I think that if we symbolically determine if certain elements of the arguments to addition are equal, then we know that broadcasting did not happen. But this required two passes. One to determine the information about the arguments and another to relate the arguments to the node type.

File: test/fx/test_gradual_type.py [hunk 9]
```diff
@@ -819,6 +823,53 @@ def test_resnet50(self):
         for n1, n2 in zip(gm_static_with_types.graph.nodes, gm_run.graph.nodes):
             assert n1.type == TensorType(n2.meta['tensor_meta'].shape)
 
+        # apply shape inference to graph and check
+        # that the batch size is equal across all layers
+        infer_symbolic_types(gm_static)
+        infer_symbolic_types(gm_static)
```
Comment (eellison) [inline test/fx/test_gradual_type.py:829]: you mentioned elsewhere that you need to run it twice - should the be invoked internally as part of `infer_symbolic_types` instead of being required at the callsite ?

File: test/fx/test_gradual_type.py [hunk 10]
```diff
@@ -819,6 +823,53 @@ def test_resnet50(self):
...
+
+
+        batch_sizes = []
```
Comment (eellison) [inline test/fx/test_gradual_type.py:832]: nit: not really a reason to not use set here from the start

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 11]
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
 
         else:
             raise NotImplementedError("Method not yet implemented")
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:528]: fyi: Linear has broadcasting behavior because it dispatches to matmul https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/Linear.cpp.html#37

Not sure if `Linear` the module is different but:

```
>>> r = torch.rand
>>> torch._C._nn.linear(r(4), r(8, 4), r(8)).size()
torch.Size([8])
```

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 12]
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
...
+    if isinstance(n.args[0], Node) and isinstance(n.args[1], Node):
+        arg_type1 = n.args[0].type
+        arg_type2 = n.args[1].type
+        if isinstance(arg_type1, TensorType) and isinstance(arg_type2, TensorType) and isinstance(n.type, TensorType):
+            args1 = arg_type1.__args__
+            args2 = arg_type2.__args__
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
+                if x == y:
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:564]: I think it would be easy enough to add a check for the `1`  behavior of broadcasting

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 13]
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
...
+@register_refinement_rule(torch.flatten)
+def flatten_refinement_rule(n: Node):
+    assert isinstance(n.args[0], Node)
+
+    eq_const = []
+
+    start_dim = 1
+    end_dim = -1
+
+    if len(n.args) > 1:
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:589]: would it be easier to assume that the arguments are normalized before this ? i dont know if this handles things like kwargs

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 14]
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
...
+    if isinstance(n.type, TensorType) and isinstance(n.args[0].type, TensorType):
+        l = len(n.type.__args__)
+        start_dim = l if start_dim == -1 else start_dim
+        end_dim = l + end_dim + 1 if end_dim < 0 else end_dim + 1
+
+        for t1, t2 in zip(n.type.__args__[0:start_dim], n.args[0].type.__args__[0:start_dim]):
+            eq_const.append(Equality(t1, t2))
+
+        for t1, t2 in zip(n.type.__args__[start_dim:end_dim], n.args[0].type.__args__[start_dim:end_dim]):
+            eq_const.append(Equality(t1, t2))
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:606]: Is this constraint right ? the output will not have the same number of dimensions as the input. 

```
>>> x = torch.rand([2, 3, 4, 5])
>>> torch.flatten(x, start_dim=1, end_dim=3).size()
torch.Size([2, 60])
```
dims 1 and 2 (iterating from `[1:3]`) will not have the same dimensions for the output and input

**Review by eellison — COMMENTED** [review]
looks good, a few comments

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 2 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:21]: yea you're right, I got rid of it

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 15]
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
...
+    res = []
+    if isinstance(n.args[0], Node) and isinstance(n.args[1], Node):
+        arg_type1 = n.args[0].type
+        arg_type2 = n.args[1].type
+        if isinstance(arg_type1, TensorType) and isinstance(arg_type2, TensorType) and isinstance(n.type, TensorType):
+            args1 = arg_type1.__args__
+            args2 = arg_type2.__args__
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:563]: since we type-checked the input first, if both operants and n.type are all tensors, then we know that the length must be equal. I think I should eventually put everything in one function instead of calling symbolic inference twice and calling the type-checker once.

File: test/fx/test_gradual_type.py (hunk 9 above)
Comment (migeed-z) [reply test/fx/test_gradual_type.py:829]: yeah I will do that. The tests are calling infer_symbolic_types twice. I'll change that to a single call.

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 12 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:564]: this part is handled in the type checker already! the definitions are here https://github.com/pytorch/pytorch/blob/faf82478fec1d777f1120ee2d1e06f01a1bf3325/torch/fx/experimental/graph_gradual_typechecker.py#L45

by the time we reach the symbolic inference part, we already did the best we can with any concrete types we had earlier.

Comment (eellison) [reply torch/fx/experimental/graph_gradual_typechecker.py:574]: i guess the case i'm talking about is if one value is `1` the output will be the symbolic shape of the other

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 11 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:536]: I will check this! I went based on the documentation here https://pytorch.org/docs/stable/generated/torch.nn.Linear.html which talked about the input and output shapes.

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 12 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:574]: I added a test case for this here https://github.com/pytorch/pytorch/pull/61776/files#diff-8ccda303dbd4c687535083b21b946c204650cb45561330e48b56a2299ab92752R922

is that along the lines of what you're referring to?

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 14 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:616]: This part should be fixed now

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 11 above)
Comment (eellison) [reply torch/fx/experimental/graph_gradual_typechecker.py:537]: As you've written it I think is fine for `nn.Linear` the Module..

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 16]
```diff
@@ -500,3 +531,156 @@ def type_check_node(self, n: Node):
...
+    if isinstance(n.type, TensorType) and isinstance(n.args[0].type, TensorType):
+        l = len(n.type.__args__)
+        start_dim = l if start_dim == -1 else start_dim
+        end_dim = l + end_dim if end_dim < 0 else end_dim
+
+        for t1, t2 in zip(n.type.__args__[0:start_dim], n.args[0].type.__args__[0:start_dim]):
+            eq_const.append(Equality(t1, t2))
+
+        for t1, t2 in zip(n.type.__args__[start_dim:end_dim], n.args[0].type.__args__[start_dim:end_dim]):
+            eq_const.append(Equality(t1, t2))
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:617]: i think the types from start_dim:end_dim are different between input/output

**Review by eellison — COMMENTED** [review]
lgtm, just a couple comments

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 16 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py:617]: I thought we had 3 parts to this. So from 0 to start and from start to end. Then start to end would be flattened but 0 to start and start to end would be the same dimensions?

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

**Review by jamesr66a — APPROVED** [review]
Looks good to me, cc @eellison for any final comments

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).
